librad/src/git/types/remote.rs

```diff
@@ -0,0 +1,164 @@
+    ///         types::{remote::Remote, FlatRef, Force, NamespacedRef},
+    ///     },
+    ///     hash::Hash,
+    ///     uri::{Path, Protocol, RadUrn},
+    /// };
+    ///
+    /// let id = Hash::hash(b"geez");
+    ///
+    /// // The RadUrn pointing some project
+    /// let urn = RadUrn::new(
+    ///     id.clone(),
+    ///     Protocol::Git,
+    ///     Path::parse("").unwrap(),
+    /// );
+    ///
+    /// // The working copy heads, i.e. `refs/heads/*`.
+    /// let working_copy_heads: FlatRef<String, _> = FlatRef::heads(PhantomData, None);
+    ///
+    /// // The monorepo heads, i.e. `refs/namespaces/<id>/refs/heads/*`.
+    /// let monorepo_heads = NamespacedRef::heads(id, None);
+    ///
+    /// // Setup the fetch and push refspecs.
+    /// let fetch = working_copy_heads
+    ///         .clone()
+    ///         .refspec(monorepo_heads.clone(), Force::True)
+    ///         .into_dyn();
+    /// let push = monorepo_heads.refspec(working_copy_heads, Force::False).into_dyn();
+    ///
+    /// // We point the remote to `LocalUrl` which will be of the form `rad://<id>.git`.
+    /// let url: LocalUrl = urn.into();
+    ///
+    /// // Setup the `Remote`.
+    /// let mut remote = Remote::rad_remote(url, fetch);
+    /// remote.add_pushes(vec![push].into_iter());
+    /// ```
+    pub fn rad_remote(url: Url, fetch_spec: Box<dyn AsRefspec>) -> Self {
+        Self {
+            url,
+            name: "rad".to_string(),
+            fetch_spec,
+            push_specs: vec![],
+        }
+    }
+
+    /// Add a series of push specs to the remote.
+    pub fn add_pushes<I>(&mut self, specs: I)
+    where
+        I: Iterator<Item = Box<dyn AsRefspec>>,
+    {
+        for spec in specs {
+            self.push_specs.push(spec)
+        }
+    }
+
+    /// Create the [`git2::Remote`] and add the specs.
+    pub fn create<'a>(&self, repo: &'a git2::Repository) -> Result<git2::Remote<'a>, git2::Error>
+    where
+        Url: ToString,
+    {
+        let _ = repo.remote_with_fetch(
+            &self.name,
+            &self.url.to_string(),
+            &self.fetch_spec.as_refspec(),
+        )?;
+
+        for spec in self.push_specs.iter() {
+            repo.remote_add_push(&self.name, &spec.as_refspec())?;
+        }
+
+        // To ensure that the push spec is persisted we need to call `find_remote` here.
+        // Otherwise, `remote_add_push` doesn't affect the "loaded remotes".
+        repo.find_remote("rad")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{io, marker::PhantomData};
+
+    use super::*;
+    use crate::{git::types::*, hash::Hash};
+    use librad_test::tempdir::WithTmpDir;
+
+    #[test]
+    fn can_create_remote() {
+        WithTmpDir::new::<_, io::Error>(|path| {
+            let repo = git2::Repository::init(path).expect("failed to init repo");
+
+            let id = Hash::hash(b"geez");
+            let heads: FlatRef<String, _> = FlatRef::heads(PhantomData, None);
```

Comment:
It is a bit weird having to pass a `PhantomData` -- is there no way to `impl` around this? Or maybe create a type alias at least?

Reply:
Ya, I couldn't think of a way to get around this without repeating code. Like we could do:
```rust
impl FlatRef<Remote> {
  pub fn heads(/* yada yada */) -> Self {
    Reference::heads(PhantomData, remote)
  }
}
```

But it's verrrrrry boilerplatey... Macros? :scream_cat:

Reply:
> Or maybe create a type alias at least?

I thought about that, initially and had:
```rust
type NoNamespace = PhantomData<!>;
```

But didn't know where to go from there. We'd still need to pass in `PhantomData`.

Reply:
I'll create a follow-up ticket because I don't wanna spend more time shaving yaks :sweat_smile: